공지사항 조회 API 개발

namdongho opened the pull request:
## 작업 내용(구체적으로)
- 10개씩 페이지 별로 공지사항 조회 API
- ID 별로 공지사항 조회 API

## Api
- [x] 공지사항 페이지별 조회 API
- [x] 공지사항 ID별 조회 API

## Test Result(테스트 결과 이미지 첨부)
![image](https://user-images.githubusercontent.com/51500359/110438380-a8139600-80f9-11eb-9192-08c0ddc51dfc.png)
  - 페이지별로 조회_성공

![image](https://user-images.githubusercontent.com/25305130/110570838-ee203680-8199-11eb-82a6-0a80038cc08e.png)
  - 페이지별로 조회_실패 : 1보다 작은 Size입력

![image](https://user-images.githubusercontent.com/25305130/110570920-1445d680-819a-11eb-9484-f74d8f94d87b.png)
  - 페이지별로 조회_실패 : 1보다 작은 Page 입렵

![image](https://user-images.githubusercontent.com/25305130/110448953-085c0500-8105-11eb-8178-04af18c5ad3d.png)
  - 아이디별 조회_성공

![image](https://user-images.githubusercontent.com/25305130/110448786-dcd91a80-8104-11eb-9dcd-658f245d66b5.png)
  - 아이디별 조회_실패 : 존재하지 않는 ID의 공지사항 조회

hamryt committed:
공지사항 create api 개발

hamryt committed:
공지사항 create api 개발

hamryt committed:
SecurityConfig 설정 수정

hamryt committed:
CreateNotice 테스트 메소드 이름 구체적으로 변경

코드 스타일 수정

hamryt committed:
Notice 조회 API

hamryt committed:
SecurityConfig

hamryt committed:
SecurityConfig

hamryt committed:
Notice 조회 API

10개씩 pagination을 적용하여 정보를 보여준다.

hamryt committed:
Notice 조회 API

10개씩 pagination을 적용하여 정보를 보여준다.

hamryt committed:
사용하지 않는 테스트 코드 삭제

chlalstjd430 commented on joljak-application/src/main/java/kr/joljak/api/notice/controller/NoticeController.java:
```diff
@@ -27,15 +28,16 @@
-    return noticeService.getAll();
+  public List<NoticeResponse> getAllByPages(@RequestParam int page,
+    @RequestParam(defaultValue = "10") int size) {
+    return noticeService.getNoticesByPageReqeust(page, size);
```
만약 사용자가 page나 size에 음수를 넣는다면? 어떻게 될까요~?

hamryt replied:
몰라 임마

chlalstjd430 replied:
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ

chlalstjd430 commented on joljak-application/src/main/java/kr/joljak/api/notice/controller/NoticeController.java:
```diff
@@ -27,15 +28,16 @@
   @ResponseStatus(HttpStatus.OK)
-  public List<NoticeResponse> getAll() {
-    return noticeService.getAll();
+  public List<NoticeResponse> getAllByPages(@RequestParam int page,
```
졸작 디자인보면 게시글에 작성자 이름이 있어야하는데 해당 정보가 없어서 수정해줘야 할 듯~

chlalstjd430 replied:
NoticesResponse를 따로 만들어서 List<NoticeResponse> 정보와, Page정보도 보내줘야할듯.
클라이언트에서 페이지에 대한 정보를 알아야 함!

hamryt committed:
공지사항 id별 조회 API 개발

chlalstjd430 commented on joljak-application/src/main/java/kr/joljak/api/notice/controller/NoticeController.java:
```diff
@@ -27,15 +29,23 @@
+  @ApiOperation("공지사항 개별 조회 API")
+  @GetMapping("/{id}")
+  @ResponseStatus(HttpStatus.OK)
+  public NoticeResponse getNoticeById(@PathVariable("id") Long id){
```
controller단인데 xxxById보다는 getNotice 정도가 적당한것같고, getAll() 메소드도 통일성잇게 getNotices로 변경해주는게 메소드의 역할을 읽기 쉬울듯

hamryt replied:
넹넹

chlalstjd430 commented on joljak-application/src/main/java/kr/joljak/api/notice/service/NoticeService.java:
```diff
@@ -40,10 +43,22 @@ private NoticeResponse getNoticeResponse(SimpleNotice simpleNotice) {
   }
+
+  private List<Notice> fetchPages(int page, int size) {
+    PageRequest pageRequest = PageRequest.of(page, size);
```
이렇게 내려주면 정렬은 어떤순서로될까?

hamryt replied:
swagger에서 순서대로 나오던데??

chlalstjd430 commented on joljak-application/src/main/java/kr/joljak/api/notice/service/NoticeService.java:
```diff
@@ -40,10 +43,22 @@ private NoticeResponse getNoticeResponse(SimpleNotice simpleNotice) {
+
+  public NoticeResponse getNoticeById(Long id) {
+    Notice notice = noticeRepository.findById(id)
+      .orElseThrow(() -> new NoNoticeWithSelectedIdException());
```
exception 네이밍도 기존의 코드 네이밍과 최대한 맞춰줘, 
그리고 애초에 NoNoticeWithSelectedIdException라는 네이밍이 읽기 불편하긴함..

chlalstjd430 commented on joljak-application/src/test/java/kr/joljak/api/notice/controller/GetNotices.java:
```diff
@@ -0,0 +1,76 @@
+
+  @Test
+  @WithMockUser(username = "testUser1", roles = "User")
+  public void get_Notice_By_Pages() throws Exception {
```
테스트 코드 메소드명도 기존 코드와 통일해줘,
메소드명이 getNotices면 getNotices_Success 이런식으로 아래 테스트코드들도 마찬가지

chlalstjd430 commented on joljak-application/src/test/java/kr/joljak/api/notice/controller/GetNotices.java:
```diff
@@ -0,0 +1,76 @@
+
+  @Test
+  @WithMockUser(username = "testUser1", roles = "User")
+  public void get_Notice_By_id_Success() throws Exception {
```
우리 메소드별로 클래스 분리해서 하기로하지않았니?
getNotice는 GetNotice 클래스 만들어서 진행해줘~

chlalstjd430 commented on joljak-application/src/test/java/kr/joljak/api/notice/controller/GetNotices.java:
```diff
@@ -0,0 +1,76 @@
+import org.springframework.security.test.context.support.WithMockUser;
+import org.springframework.test.web.servlet.MvcResult;
+
+public class GetNotices extends CommonApiTest {
```
과연 notice 리스트를 불러오는 상황을 실패할 경우가 없을까?
page가 음수라면?

chlalstjd430 commented on joljak-application/src/test/java/kr/joljak/api/notice/controller/GetNotices.java:
```diff
@@ -0,0 +1,76 @@
+
+  @Test
+  @WithMockUser(username = "testUser1", roles = "User")
+  public void get_Notice_By_id_Fail() throws Exception {
```
[메소드명]_[성공/실패]_[실패시 실패 사유(excpetion message)] 이렇게 네이밍하기로 정했고, 기존 코드도 그런데,, 최대한 모든 코드는 기존 코드와 일관성을 갖쳐줘~

hamryt replied:
엉 바꿀게 지금 알았넹

chlalstjd430 commented on joljak-domain-rds/src/main/java/kr/joljak/domain/user/exception/NoNoticeWithSelectedIdException.java:
```diff
@@ -0,0 +1,12 @@
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(value = HttpStatus.NOT_FOUND)
+public class NoNoticeWithSelectedIdException extends RuntimeException {
```
왜 기존 코드대로 안하는거야..?
기존 코드 스타일 참고해서 변경해줘

chlalstjd430 reviewed (commented):
지금 코드 작성시 기존 코드와 일관되지않게 작성하는데, 이거는 클린 코드 관점에서도 가독성을 방해할 뿐더러, 추후 지속적인 개발을 할 때도 일관되않는 코드를 작성하여 연속적으로 혼동을 야기해

최대한 일관성있게 코드 작성해줘~

chlalstjd430 reviewed (commented):
@hamryt 
생각해보니 NoticeService application단이 아니라 domain 계층에 있는게 맞는 것 같아, 가급적이면 repository를 멤버로 두고 있는 서비스는 도메인 계층에서 만들자!

이 부분 조금 번거롭겠지만 수정해줄 수 있을까?

hamryt committed:
Test코드 스타일 수정
NoticesResponse 생성: 페이지 별로 받을 때 페이지 정보랑 List 받기 위해
Controller 메소드 이름 수정
Service 메소드 이름 수정
NoticeNotFoundException 예외 이름 수정

chlalstjd430 commented on joljak-application/src/main/java/kr/joljak/api/notice/response/NoticesResponse.java:
```diff
@@ -0,0 +1,21 @@
+public class NoticesResponse {
+
+  private List<NoticeResponse> noticeResponseList;
+  private int page;
```
page(int)가 아닌 pageable객체를 보내줘~

hamryt replied:
예아~

chlalstjd430 commented on joljak-application/src/main/java/kr/joljak/api/notice/service/NoticeService.java:
```diff
@@ -36,29 +37,37 @@ private NoticeResponse getNoticeResponse(SimpleNotice simpleNotice) {
+  @Transactional(readOnly = true)
+  public NoticesResponse getNoticesByPage(int page, int size) {
+    List<Notice> noticeList;
+    if (page < 0) {
```
공통코드로 분리하기로 하지않았나?

chlalstjd430 commented on joljak-application/src/test/java/kr/joljak/api/notice/controller/GetNotices.java:
```diff
@@ -15,12 +15,9 @@
   @Test
   @WithMockUser(username = "testUser1", roles = "User")
-  public void get_Notice_By_Pages() throws Exception {
+  public void getNotices_Success() throws Exception {
```
페이지가 음수인 경우에도 잘 되는지 테스트 코드 추가해주는게 좋을듯!

chlalstjd430 commented on joljak-domain-rds/src/main/java/kr/joljak/domain/util/FetchPages.java:
```diff
@@ -0,0 +1,14 @@
+  private FetchPages() {
+  }
+
+  public static final PageRequest of(int page, int size){
```
여기서 page,size가 음수라면 default값(page = 0, size = 10)로 전환해주는 코드를 넣으면될듯!

chlalstjd430 commented on joljak-domain-rds/src/main/java/kr/joljak/domain/util/FetchPages.java:
```diff
@@ -0,0 +1,14 @@
+
+public class FetchPages {
+
+  private FetchPages() {
```
생성자를 private로 해준건 정말 잘한거같아!

chlalstjd430 reviewed (commented):
리뷰 준 내용 다시한번 반영 부탁할게~

hamryt committed:
Page변수 Pageable 객체 타입으로 변경
page 초기화 로직 FetchPages클래스 안으로 이동
page, size 변수 예외 상황에 대한 테스트 코드 추가

chlalstjd430 commented on joljak-application/src/test/java/kr/joljak/api/notice/controller/GetNotices.java:
```diff
@@ -27,4 +27,28 @@ public void getNotices_Success() throws Exception {
 
+  @Test
+  @WithMockUser(username = "testUser1", roles = "User")
+  public void getNotices_Fail_SizeUnderZero() throws Exception {
```
getNotices_Success로 아래 메소드까지 변경해주면될 것 같아!

hamryt replied:
엉키엉키

chlalstjd430 reviewed (dismissed):
코등 승인할게!
그 테스트 코드 메소드명만 바꾸고 리뷰없이 그냥 머지해줘~!
수고했다~

hamryt committed:
테스트 메소드 이름 변경
